hw10-autologger

SGrechnev opened the pull request:
'./gradlew run' for run Demo.main()

SGrechnev committed:
hw10-autologger
'./gradlew run' for run Demo.main()

spvgl commented on hw10-autologger/src/main/java/ru/otus/TestLoggingImpl.java:
```diff
@@ -0,0 +1,31 @@
+public class TestLoggingImpl implements TestLogging {
+    @Log
+    public void calculation(int param) {
+//        System.out.println("\tTestLogging.calculation(Integer)");
```
А зачем закоментировали здесь и ниже?
Наоборот, оставьте, чтобы было видно, что метод вызывается.

SGrechnev replied:
Исправил

spvgl commented on hw10-autologger/src/main/java/ru/otus/logger/Logger.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public static TestLogging createTestLogging() {
+        InvocationHandler handler = new DemoInvocationHandler(new TestLoggingImpl());
```
Давайте сделаем универсальный `TestLogging` и уберем из этого класса захардкоженый `TestLoggingImpl` и `TestLogging`, чтобы можно был работать с любом объектом (может с какими-то ограничениями).

Объект (`TestLoggingImpl`) создаем снаружи, передаем его сюда  и получаем прокси, который логирует.

Для проверки добавьте еще какой-нибудь класс, помимо `TestLoggingImpl`.

SGrechnev replied:
Согласен, очень логично.
В таком случае аннотации @Log должны находиться в интерфейсе, правильно? В примере из формулировки ДЗ аннотацией помечаются методы класса, реализующего интерфейс.

spvgl replied:
> В таком случае аннотации @log должны находиться в интерфейсе, правильно?

Нет. Аннотации должны остаться в классе. 
Это еще одна задачка, но вполне решаемая.

SGrechnev replied:
Аннотации оставил в классах.
В TestLoggingImpl1 и TestLoggingImpl2 аннотации расставил по-разному.

spvgl commented on hw10-autologger/src/main/java/ru/otus/logger/Logger.java:
```diff
@@ -0,0 +1,51 @@
+            ).toArray(Method[]::new);
+
+            if (foundMethods.length != 1) {
+                // assumed unreachable
+                throw new IllegalStateException("There are " + foundMethods.length + " methods" +
+                        " with equal name and param types: " + Arrays.toString(foundMethods));
```
Поясните, пожалуйста, что это и для чего?

SGrechnev replied:
Теоретически, мы не найдем два или более метода с одинаковым именем и типами параметров. Также вряд ли возможна ситуация, в которой мы не нашли метода с указанной сигнатурой, ведь объект должен реализовывать интерфейс TestLogging.
Всё же, сделал проверку, что найден ровно один метод с нужной сигнатурой.

spvgl replied:
> Теоретически, мы не найдем два или более метода с одинаковым именем и типами параметров.

Я бы сказал и практически не найдем.
Лучше убрать эту проверку. Она лишняя и только запутывает и возникают вопросы.

SGrechnev replied:
Хорошо.

spvgl commented on hw10-autologger/src/main/java/ru/otus/logger/Logger.java:
```diff
@@ -0,0 +1,51 @@
+                        " with equal name and param types: " + Arrays.toString(foundMethods));
+            }
+
+            if (foundMethods[0].isAnnotationPresent(Log.class)) {
```
Можно в начале найти сразу все методы с `@Log`, т.е. те, которые надо логировать, и  запомнить их. 
В `invoke()` проверять есть ли среди методов, которые запомнили, вызываемый метод.

SGrechnev committed:
Universal Logger for TestLogging
https://github.com/SGrechnev/2020-09-otus-java-grechnev/pull/5#discussion_r521970922

spvgl commented on hw10-autologger/src/main/java/ru/otus/logger/Logger.java:
```diff
@@ -0,0 +1,48 @@
+        private final TestLogging testLogging;
+        Method[] loggingMethods;
+
+        DemoInvocationHandler(TestLogging testLogging) {
```
```suggestion
        DemoInvocationHandler(TestLogging obj) {
```

spvgl commented on hw10-autologger/src/main/java/ru/otus/logger/Logger.java:
```diff
@@ -0,0 +1,48 @@
+
+    static class DemoInvocationHandler implements InvocationHandler {
+        private final TestLogging testLogging;
+        Method[] loggingMethods;
```
```suggestion
        private Method[] loggingMethods;
```

spvgl commented on hw10-autologger/src/main/java/ru/otus/logger/Logger.java:
```diff
@@ -0,0 +1,48 @@
+                    break;
+                }
+            }
+            if (i < this.loggingMethods.length) {
```
Это работает, но как-то совсем не читаемо.
Лучше сделать явную переменную `boolean needLog`,  установить ее в цикле и потом проверить.

Перебирать тоже не очень оптимально - можно запомнить сразу имя метода с его параметрами в виде строки. Потом для вызываемого метода вычислить этот "ключ" (имя + параметры) и сравнить его с запомненными.

spvgl commented on hw10-autologger/src/main/java/ru/otus/logger/Logger.java:
```diff
@@ -0,0 +1,48 @@
+    private Logger() {
+    }
+
+    public static TestLogging createTestLogging(TestLogging proxiedObject) {
```
Хотелось бы совсем уйти от зависимостей от интерфейсов (TestLogging) в Logger.

Чтобы можно было передать сюда класс и с другим интерфейсом.
